Second batch of improvements to Carousel

sgomes opened the pull request:
This is the second batch of improvements to carousel, building upon #18858 and more recent changes such as #19843.

This batch of improvements focuses on reorganising the codebase and preparing for future work that removes the dependency on `jQuery`.

#### Changes proposed in this Pull Request:
- Stop implementing the carousel as a `jQuery` extension.
  Beyond locking us into `jQuery`, this was unnecessary and confusing, given that there is only ever one instance of a carousel on the page.
- Reimplement all jQuery extension functionality as plain old functions.
- Reimplement touchwipe as a plain function
- Rename functions to follow proper naming conventions and be more descriptive of their behaviour.
- Use a flag to enable and disable keyboard bindings instead of binding and unbinding the listeners all the time.
- Standardise variables on camelCase instead of snake_case.
- Break up carousel initialization into more manageable pieces.
- Move away from PHP-style Yoda conditions.
- Stop using `var foo = function()` syntax everywhere.
- Simplify some logic and remove some duplication.
- `use strict`, to make sure we catch missing declarations and the like.
- Remove commented-out and dead code.
- Some style fixes.

#### Jetpack product discussion
For a high-level, somewhat outdated plan, see p1HpG7-bfN-p2.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Smoke-test the carousel. There should be no changes to functionality, as all of the modifications were focused on improving the codebase, with no intended user-visible changes.

sgomes commented:
Added a final change in 16ea7e2cd157e20986c0ea6cd44b8c81f0f0ac7b.

sgomes commented:
Marking as not ready for review while investigating some issues.

sgomes commented:
This should again be good to review.

sdixon194 commented:
Nice! What a refactor.

Did a good bit of testing on this one again. 

Looks like the only big thing I noticed was that the button style changed:

before:

![Screen Shot on 2021-05-18 at 13:36:47](https://user-images.githubusercontent.com/33553323/118699376-7432ac00-b7df-11eb-91d2-e974ff1c453d.png)

after:

![Screen Shot on 2021-05-18 at 13:37:36](https://user-images.githubusercontent.com/33553323/118699389-77c63300-b7df-11eb-885d-82c4d83f5c77.png)

Maybe we should get design input if we want to see about updating the styling, if that's something that's going to happen as a result of these improvements? 

## Non Blocking
These were issues I noticed occurred with the gallery, but seems to also happen on latest stable, so I don't believe they were introduced with this PR.

### Prior to 9.7

The following two also happened on Jetpack 9.5, so seem unlikely related to previous improvements and I could open separate issues if necessary:

**Long Description Doesn't Wrap**

Images with long a long, unbroken string of characters in the description (like a URL), doesn't wrap):
![Screen Shot on 2021-05-18 at 13:46:59](https://user-images.githubusercontent.com/33553323/118699448-8ca2c680-b7df-11eb-8133-42d1a2e44520.png)

**Meta data doesn't collapse nicely**

Related to above, when shrinking the gallery but before the breakpoint, the meta area gets squished when the image has a description:

![Screen Shot on 2021-05-18 at 13:42:22](https://user-images.githubusercontent.com/33553323/118699558-a9d79500-b7df-11eb-9d34-9048d2335fd0.png)

### After 9.7
This one doesn't occur prior to 9.7, so likely introduced in #18858

**Meta data stuck to bottom when no comment**
Seems the meta data is stuck to the bottom without padding, [similar to this issue](https://github.com/Automattic/jetpack/pull/18858#issuecomment-805963164).

![Screen Capture on 2021-05-18 at 13-35-56](https://user-images.githubusercontent.com/33553323/118700347-8fea8200-b7e0-11eb-929c-8b362b87ce1e.gif)

Prior to 9.7:
![Screen Shot on 2021-05-18 at 14:01:29](https://user-images.githubusercontent.com/33553323/118701317-93323d80-b7e1-11eb-9211-c88c0b5b4317.png)

jeherve commented:
> Maybe we should get design input if we want to see about updating the styling, if that's something that's going to happen as a result of these improvements?

I personally like the new style better, it better fits the rest of the site since it uses styles from the theme, but I suppose that for some themes, it may actually make things less readable.

Let's see what folks on the Design team think about this :)

folletto commented:
I think the style should match the dark overlay, thus not inheriting neither from the system default, nor from the theme, as the dark overlay doesn't inherit from either.

We can thus match the colors to the rest of the black/grays/white:

<img width="658" alt="Screenshot 2021-05-19 at 09 21 04" src="https://user-images.githubusercontent.com/4389/118780080-98e05f80-b883-11eb-9368-0dc09f99121c.png">

* border white, text white
* background to match the light background of the textarea

keoshi commented:
I echo @folletto's suggestion but also noting that we do support a light theme in addition to the default dark one:

![image](https://user-images.githubusercontent.com/390760/118782333-e8c02600-b885-11eb-9c2d-24bcfe0df0ca.png)

![image](https://user-images.githubusercontent.com/390760/118782382-f37abb00-b885-11eb-9c17-7cacd0c4589a.png)

Here's how that one could look like:

![image](https://user-images.githubusercontent.com/390760/118782900-769c1100-b886-11eb-8f7b-b2687d09f7a9.png)

I followed the same guidelines as Erin, only adjusting the font-size to 14px instead of 13px (size in “View full size”), and using `8px 16px` for the button's padding.

sgomes commented:
Thank you for the extensive review, @sdixon194, and thank you everyone else for the design feedback!

I have to confess that the change is unintentional (and I'm not entirely sure whether it's related to this particular PR or any previous ones), but I'm more than happy to incorporate the feedback 👍 

I'll update the labels once I've folded that in.

sdixon194 commented:
> I have to confess that the button change is unintentional (and I'm not entirely sure whether it's related to this particular PR or any previous ones), but I'm more than happy to incorporate the feedback

Either way, looks great! 😂 

Did some more testing and does look like it occurs when standing on master prior to applying this branch, so likely something else that was merged in this cycle? I could do some more digging on that, but either way then nothing seems to be a particular blocker for this PR unless you want to make some of those tweaks!

samiff commented:
Hello @sgomes 👋 

Were you planning additional changes for this PR? Next code freeze is on May 24, 2021 if it helps in planning.

sgomes commented:
Thank you for checking in, @samiff! 👍 

I'm planning on getting the suggested improvements added to this PR today. If for some reason I can't get them ready, however, I'll update the thread, since it seems that the improvements are non-critical and hopefully we could get the PR approved as-is.

Give me a few hours and I'll let you know how it goes! 😄

sgomes commented:
All of the issues mentioned above should now be fixed, so this is ready for a final review!

sgomes commented:
Rebased, to please one of the PR checks.

sdixon194 commented:
Heck yeah, thanks for addressing those issues @sgomes ! This works super well for me. The only issue above that still occurs for me now is the long unbroken description not wrapping. Was that intended as part of the fixes you mentioned working on, or was that something we won't address for this PR? 

Love how the photo meta data works on resize now. So so so much cleaner.

sgomes commented:
> Heck yeah, thanks for addressing those issues @sgomes ! This works super well for me. 

Woohoo, that's great! 😄

> The only issue above that still occurs for me now is the long unbroken description not wrapping. Was that intended as part of the fixes you mentioned working on, or was that something we won't address for this PR?

Oh, my solution to the wrapping problem was to use an ellipsis whenever a word got too long, through `text-overflow: ellipsis;`.

![image](https://user-images.githubusercontent.com/409615/119166177-3be8d300-ba56-11eb-9e55-37d82a91adf0.png)


It sounds like you're looking for something like `overflow-wrap: break-word` instead?

![image](https://user-images.githubusercontent.com/409615/119166270-558a1a80-ba56-11eb-8748-ff3e8143a010.png)

I think both are valid options, and it's a one-line CSS change; let me know which one you prefer.

> Love how the photo meta data works on resize now. So so so much cleaner.

\o/

sdixon194 commented:
Ah strange, on my test site I don't see the ellipses: https://motionless-barnacle.jurassic.ninja/2021/05/21/gallery-test/

Tried a few different browsers and even adding the CSS in the customizer, but couldn't get it to work 🤔 

I was able to get `overflow-wrap: break-word` to work though. I think that also might be the better solution, since it ensures all the text is visible to cover all outlier cases, but open to being persuaded otherwise!

sgomes commented:
Interesting, there does seem to be something about your test site that prevents `text-overflow: ellipsis` from working. It could be something related to the theme.

> I was able to get `overflow-wrap: break-word` to work though. I think that also might be the better solution, since it ensures all the text is visible to cover all outlier cases, but open to being persuaded otherwise!

In any case, since breaking the word into multiple lines seems preferable, I went ahead and switched to that approach 🙂

Thank you for all the excellent testing, @sdixon194! 👍

sgomes commented:
Oh, I figured it out; it's because of the `<p>` tags inside the description, which my hacky testing approach was removing 😄

In any case, breaking the word seems to work everywhere, so we should be good 👍

sdixon194 reviewed (approved):
Excellent, works for me now! 👍 🎉

sgomes commented:
Woohoo, thanks again, @sdixon194! 🙇 

@samiff: Is it ok to leave the merge for Monday, so that I can follow-up with the wpcom deploy straight away? Monday's the deadline you mentioned, but I'm not sure whether it's inclusive or exclusive 😄 

Otherwise, please feel free to merge before then, and I'll handle the wpcom side of things on Monday!

samiff commented:
> Is it ok to leave the merge for Monday

@sgomes Yes, merging this on Monday will work 👍

sgomes commented:
Thanks again, everyone, merged! 👍

sgomes commented:
r226125-wpcom
